### ISSUE TITLE
Long lived dropped guards.

### ISSUE THREAD

**Issue by BruceBrown** [Yoric/timer.rs#28]
Dropping a guard prevents timers from firing. I've a case where I end up with 1000s of timers that are hours in the future that have had there guard dropped. Consider adding a periodic sweep through the scheduled timers to remove entries that have had their guard dropped.

There is likely a good place to call `self.heap.retain(|s| s.guard.should_execute());`  within the scheduler loop. Ideally, not every time through the loop.

**Comment by Yoric** [comment]
Oh, interesting idea!

**Comment by Yoric** [comment]
So, do I understand correctly that the problem you're pointing out is that the timers retain resources even after being dropped?

**Comment by BruceBrown** [comment]
Yes. Dropping the guard doesn't clean up the Schedules in the Scheduler; it just changes the state to not execute when the timer expires. In my case I have 1000s of events scheduled for hours into the future Very few of them will ever fire, as most will be dropped when state changes. However, even though I may only have a few that remain active, the scheduler has all that have yet to expire.  Its that bloat that I'd hope to see fixed. I could prepare a PR, but I don't like where I'd place the change, and you're likely to know of a better place for it.
